src/client/datascience/jupyter/kernels/kernelProvider.ts

```diff
@@ -42,8 +42,21 @@ export class KernelProvider implements IKernelProvider {
     }
     public getOrCreate(uri: Uri, options: KernelOptions): IKernel | undefined {
         const existingKernelInfo = this.kernelsByUri.get(uri.toString());
-        if (existingKernelInfo && fastDeepEqual(existingKernelInfo.options.metadata, options.metadata)) {
-            return existingKernelInfo.kernel;
+        if (existingKernelInfo) {
```

Comment:
This getOrCreate runs on every cell execution. It's checking to see if the new KernelOptions specified still match the current kernel running the notebook. If it matches it just keeps using that kernel, if it doesn't match it starts a new one.

When you launch via a kernelspec that doesn't have an associated interpreter it just fills in the interpreter in the metadata with the current active interpreter. So if you change the active interpreter you can now create a mismatch here and the kernel restarts.

Reply:
Essentially, if you are doing kernelspec launches, then to compare if you are already running the specified kernelspec just look at the kernelspec, not the interpreter that you launched or would launch it with.

Reply:
> When you launch via a kernelspec that doesn't have an associated interpreter it just fills in the interpreter in the metadata with the current active interpreter. So if you change the active interpreter you can now create a mismatch here and the kernel restarts.

Got it, thanks
I don' think my suggestion works in this case.

Reply:
Added a bug for https://github.com/microsoft/vscode-jupyter/issues/4194 where it doesn't work in this case. Ideally we should be able to use that, if not then the function is not doing what its meant to do.